### PR TITLE
fix(pymongo): use bytecode wrapping to trace pymongo clients

### PR DESCRIPTION
**Pull request by mabdinur**
Resolves: https://github.com/DataDog/dd-trace-py/issues/10344

- Uses bytecode wrapping instead of wrapt's objectproxys to instrument mongo clients and socket connections
- Previously the pin object used in the MongoClient was different from the ones used to trace socket connections.  Updates usages of the Pin object to ensure sever, socket, and client objects use the same Pin. This sublte fix required updating the span counts in tests (now the DummyTracer used in tests is expected to capture one more span per trace).
- Updates the mongoengine integration to simply use the pymongo integration. This integration no longer does anything special and can be removed.
- Improves the test coverage of the pymongo integration. Ensures every pymongo version range with custom patching has test coverage.

## Checklist
- [x] PR author has checked that all the criteria below are met
- The PR description includes an overview of the change
- The PR description articulates the motivation for the change
- The change includes tests OR the PR description describes a testing strategy
- The PR description notes risks associated with the change, if any
- Newly-added code is easy to change
- The change follows the [library release note guidelines](https://ddtrace.readthedocs.io/en/stable/releasenotes.html)
- The change includes or references documentation updates if necessary
- Backport labels are set (if [applicable](https://ddtrace.readthedocs.io/en/latest/contributing.html#backporting))

## Reviewer Checklist
- [x] Reviewer has checked that all the criteria below are met 
- Title is accurate
- All changes are related to the pull request's stated goal
- Avoids breaking [API](https://ddtrace.readthedocs.io/en/stable/versioning.html#interfaces) changes
- Testing strategy adequately addresses listed risks
- Newly-added code is easy to change
- Release note makes sense to a user of the library
- If necessary, author has acknowledged and discussed the performance implications of this PR as reported in the benchmarks PR comment
- Backport labels are set in a manner that is consistent with the [release branch maintenance policy](https://ddtrace.readthedocs.io/en/latest/contributing.html#backporting)

### CODE REVIEW

File: ddtrace/contrib/internal/pymongo/client.py [hunk 1]
```diff
@@ -88,6 +114,9 @@ def __getddpin__(self):
 
 @contextlib.contextmanager
 def wrapped_validate_session(wrapped, instance, args, kwargs):
+    # The function is exposed in the public API, but it is not used in the codebase.
+    # TODO: Remove this function when ddtrace.contrib.pymongo.client is removed.
```
Comment (mabdinur) [inline ddtrace/contrib/internal/pymongo/client.py:118]:
```suggestion
    # TODO(mabdinur): Remove this function when ddtrace.contrib.pymongo.client is removed.
```

Comment (mabdinur) [inline ddtrace/contrib/internal/pymongo/client.py:118]:
```suggestion
    # TODO(mabdinur): Remove this function when ddtrace.contrib.pymongo.client is removed.
```

File: releasenotes/notes/munir-pymongo-fix-patching-545f3947644c91cd.yaml [hunk 2]
```diff
@@ -0,0 +1,4 @@
+---
+fixes:
+  - |
+    pymongo: Ensures instances of the ``pymongo.MongoClient`` can be patch after pymongo is imported.
```
Comment (emmettbutler) [inline releasenotes/notes/munir-pymongo-fix-patching-545f3947644c91cd.yaml:4]:
```suggestion
    pymongo: Ensures instances of the ``pymongo.MongoClient`` can be patched after pymongo is imported.
```

File: ddtrace/contrib/internal/mongoengine/patch.py [hunk 3]
```diff
@@ -1,5 +1,9 @@
+# TODO(mabdinue): Remove the pymongoengine integration, this integration does nothing special
```
Comment (emmettbutler) [inline ddtrace/contrib/internal/mongoengine/patch.py:1]:
```suggestion
# TODO(mabdinur): Remove the mongoengine integration, this integration does nothing special
```

**Review by emmettbutler — APPROVED** [review]

File: ddtrace/contrib/internal/pymongo/client.py [hunk 4]
```diff
@@ -88,6 +114,9 @@ def __getddpin__(self):
 
 @contextlib.contextmanager
 def wrapped_validate_session(wrapped, instance, args, kwargs):
+    # The function is exposed in the public API, but it is not used in the codebase.
+    # TODO(<owner>): Remove this function when ddtrace.contrib.pymongo.client is removed.
```
Comment (mabdinur) [inline ddtrace/contrib/internal/pymongo/client.py:118]:
```suggestion
    # TODO(mabdinur): Remove this function when ddtrace.contrib.pymongo.client is removed.
```

File: ddtrace/contrib/internal/pymongo/client.py [hunk 5]
```diff
@@ -45,6 +46,31 @@
...
+        return ddtrace.Pin.get_from(client._topology)
+
+    client.__setddpin__ = functools.partial(__setddpin__, client)
+    client.__getddpin__ = functools.partial(__getddpin__, client)
+
+    # Default Pin
+    ddtrace.Pin(service=_DEFAULT_SERVICE).onto(client)
+
+
+# TODO(<owner>): Remove TracedMongoClient when ddtrace.contrib.pymongo.client is removed from the public API.
```
Comment (mabdinur) [inline ddtrace/contrib/internal/pymongo/client.py:73]:
```suggestion
# TODO(mabdinur): Remove TracedMongoClient when ddtrace.contrib.pymongo.client is removed from the public API.
```

File: ddtrace/contrib/internal/pymongo/client.py [hunk 6]
```diff
@@ -45,6 +46,31 @@
 _DEFAULT_SERVICE = schematize_service_name("pymongo")
 
 
+def trace_mongo_client_init(func, args, kwargs):
+    # Call MongoClient.__init__
```
Comment (mabdinur) [inline ddtrace/contrib/internal/pymongo/client.py:50]:
```suggestion
```

File: tests/contrib/pymongo/test.py [hunk 7]
```diff
@@ -842,4 +864,4 @@ def test_server_info(self):
 
         assert len(spans) == 2
         self.check_socket_metadata(spans[0])
-        assert spans[1].name == "pymongo.cmd"
+        assert spans[1].name == "pymongo.cmd"
```
Comment (mabdinur) [inline tests/contrib/pymongo/test.py:867]:
```suggestion
        assert spans[1].name == "pymongo.cmd"

```

**Review by quinna-h — APPROVED** [review]

File: ddtrace/contrib/internal/pymongo/client.py [hunk 8]
```diff
@@ -105,213 +91,192 @@
...
+def _trace_server_send_message_with_response(func, args, kwargs):
+    server_instance = get_argument_value(args, kwargs, 0, "self")
+    operation = get_argument_value(args, kwargs, 1, "operation")
+
+    span = _datadog_trace_operation(operation, server_instance)
+    if span is None:
+        return func(*args, **kwargs)
+    with span:
+        result = func(*args, **kwargs)
+        if result:
```
[the remark this answers is not in the document]
Comment (mabdinur) [reply ddtrace/contrib/internal/pymongo/client.py:181]: I will not refactor this code in this PR. I will keep this logic as is

File: ddtrace/contrib/internal/mongoengine/patch.py [hunk 9]
```diff
@@ -1,5 +1,9 @@
+# TODO(mabdinur): Remove the pymongoengine integration, this integration does nothing special
+# it just uses the pymongo integration and creates unnecessary pin objects
```
Comment (erikayasuda) [inline ddtrace/contrib/internal/mongoengine/patch.py:2]: nit: Do we want to mark it as deprecated then? And fully remove in the next major version EOY?

File: ddtrace/contrib/internal/pymongo/client.py [hunk 10]
```diff
@@ -45,47 +46,32 @@
 _DEFAULT_SERVICE = schematize_service_name("pymongo")
 
 
+# TODO(mabdinur): Remove TracedMongoClient when ddtrace.contrib.pymongo.client is removed from the public API.
```
Comment (erikayasuda) [inline ddtrace/contrib/internal/pymongo/client.py:49]: nit: Same question here about marking as deprecated

File: tests/contrib/mongoengine/test.py [hunk 11]
```diff
@@ -42,8 +40,9 @@ def test_insert_update_delete_query(self):
 
         # ensure we get a drop collection span
         spans = tracer.pop()
-        assert len(spans) == 1
-        span = spans[0]
+        assert len(spans) == 2
```
Comment (erikayasuda) [inline tests/contrib/mongoengine/test.py:43]: Sorry dumb question, why did this change to 2? 😅

File: tests/contrib/pymongo/test.py [hunk 12]
```diff
@@ -1,7 +1,7 @@
 # stdlib
+# 3p
```
Comment (erikayasuda) [inline tests/contrib/pymongo/test.py:2]: nit: So this only applies for `time`? Since we're using our local wrapped `pymongo`?

Comment (erikayasuda) [reply tests/contrib/pymongo/test.py:2]: Also didn't know we did this 😅

**Review by erikayasuda — APPROVED** [review]
Just a few questions for my own clarification. LGTM!

File: ddtrace/contrib/internal/mongoengine/patch.py (hunk 9 above)
Comment (mabdinur) [reply ddtrace/contrib/internal/mongoengine/patch.py:2]: This is a weird one. We want to deprecate this integrations public api (patch, unpatch, get_versions). 

We will still provide support for this integration we just don't need the code in ddtrace.contrib.internal.mongoengine to do it. I am not sure how to communicate this. Customer's shouldn't be calling patch, unpatch, or get_versions anyway. 

Maybe log a deprecation warning in `ddtrace.patch_all()` when mongoengine=True.

File: ddtrace/contrib/internal/pymongo/client.py (hunk 10 above)
Comment (mabdinur) [reply ddtrace/contrib/internal/pymongo/client.py:49]: We already have deprecation warnings for this. With @rachelyangdog's work this class will be removed from the public api in v3.0.0.

Comment (erikayasuda) [reply ddtrace/contrib/internal/pymongo/client.py:49]: Oh oops, you're right this is the internal patch.py

File: tests/contrib/mongoengine/test.py (hunk 11 above)
Comment (mabdinur) [reply tests/contrib/mongoengine/test.py:43]: I made a subtle change to the pin object. I made sure the SAME pin object is applied to client, server, and socket connections. Previously the pin object used in this test had two different tracers for client and server connections. So this test was only configured to capture and assert the number of client spans. This looked wrong to me.

Now this test captures all the spans generated by the pymongo client and server objects in one tracer.

File: tests/contrib/pymongo/test.py (hunk 12 above)
Comment (mabdinur) [reply tests/contrib/pymongo/test.py:2]: This is from linting this really old file. I didn't add this comment. We can remove it.

Comment (mabdinur) [inline tests/contrib/pymongo/test.py:2]:
```suggestion
```

File: ddtrace/contrib/internal/mongoengine/patch.py (hunk 9 above)
Comment (erikayasuda) [reply ddtrace/contrib/internal/mongoengine/patch.py:2]: Yeah that makes sense. If we've already marked the integration as deprecated on the public side, then we can probably just deprecate the internal code along with it

**Pull request merged**